File: backend/User/src/main/java/tn/esprit/user/repository/UserRepository.java
package tn.esprit.user.repository;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import tn.esprit.user.entity.User;
import java.util.Optional;
import org.springframework.data.repository.query.Param;
import tn.esprit.user.entity.UserRole;
import java.util.List;

public interface UserRepository extends JpaRepository<User, String> {
    Optional<User> findByEmail(String email);
    Optional<User> findByKeycloakId(String keycloakId);   // NEW
    boolean existsByEmail(String email);
    @Query("SELECT u FROM User u WHERE u.role = :role AND (LOWER(u.name) LIKE LOWER(CONCAT('%', :query, '%')) OR LOWER(u.email) LIKE LOWER(CONCAT('%', :query, '%')))")
    List<User> searchByRoleAndQuery(@Param("query") String query, @Param("role") UserRole role);
    List<User> findByDoctorEmail(String doctorEmail);
}

File: user-node-service/src/users/repositories/user.repository.ts
import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';
import { User, UserDocument } from '../schemas/user.schema';
import { UserRole } from '../schemas/user-role.enum';
import { IUserRepository } from '../interfaces/user-repository';

@Injectable()
export class UserRepository implements IUserRepository {
  constructor(
    @InjectModel(User.name) private readonly userModel: Model<UserDocument>,
  ) {}

  async create(userData: Partial<User>): Promise<UserDocument> {
    const user = new this.userModel(userData);
    return user.save();
  }

  async findByEmail(email: string): Promise<UserDocument | null> {
    return this.userModel.findOne({ email }).exec();
  }

  async findByKeycloakId(keycloakId: string): Promise<UserDocument | null> {
    return this.userModel.findOne({ keycloakId }).exec();
  }

  async findById(userId: string): Promise<UserDocument | null> {
    return this.userModel.findOne({ userId }).exec();
  }

  async findByRole(role: UserRole): Promise<UserDocument[]> {
    return this.userModel.find({ role }).exec();
  }

  async existsByEmail(email: string): Promise<boolean> {
    const user = await this.userModel.findOne({ email }).exec();
    return !!user;
  }

  async searchByRoleAndQuery(
    query: string,
    role: UserRole,
  ): Promise<UserDocument[]> {
    return this.userModel
      .find({
        role,
        $or: [
          { name: { $regex: query, $options: 'i' } },
          { email: { $regex: query, $options: 'i' } },
        ],
      })
      .exec();
  }

  async findAll(): Promise<UserDocument[]> {
    return this.userModel.find().exec();
  }

  async save(user: UserDocument): Promise<UserDocument> {
    return user.save();
  }

  async delete(user: UserDocument): Promise<void> {
    await user.deleteOne();
  }

  async updateMany(filter: any, update: any): Promise<any> {
    return this.userModel.updateMany(filter, update).exec();
  }
}
